contracts/finance/Vesting.sol

```diff
@@ -0,0 +1,279 @@
+        return
+            _vestingCalculation(
+                schedule_,
+                vesting_.vestingAmount,
+                vesting_.vestingStartTime,
+                timestampUpTo_,
+                block.timestamp
+            ) - vesting_.paidAmount;
+    }
+
+    // get released amount at the moment
+    function _getVestedAmount(
+        VestingData storage vesting_,
+        Schedule storage schedule_,
+        uint256 timestampUpTo_
+    ) internal view virtual returns (uint256) {
+        return
+            _vestingCalculation(
+                schedule_,
+                vesting_.vestingAmount,
+                vesting_.vestingStartTime,
+                timestampUpTo_,
+                block.timestamp
+            );
+    }
+
+    function _createVesting(VestingData memory vesting_) internal virtual returns (uint256) {
+        require(
+            vesting_.vestingStartTime > 0,
+            "VestingWallet: cannot create vesting for zero time"
+        );
+        require(
+            vesting_.vestingAmount > 0,
+            "VestingWallet: cannot create vesting for zero amount"
+        );
+        require(
+            vesting_.beneficiary != address(0),
+            "VestingWallet: cannot create vesting for zero address"
+        );
+
+        Schedule memory _schedule = _schedules[vesting_.scheduleId];
+
+        require(
+            _schedule.scheduleData.cliffInPeriods < _schedule.scheduleData.durationInPeriods,
+            "VestingWallet: cliff cannot be greater than duration"
+        );
+
+        require(
+            vesting_.vestingStartTime +
+                _schedule.scheduleData.durationInPeriods *
+                _schedule.scheduleData.secondsInPeriod >
+                block.timestamp,
+            "VestingWallet: cannot create vesting for past date"
+        );
+
+        _vestingId++;
+
+        _beneficiaryIds[vesting_.beneficiary].add(_vestingId);
+
+        _vestings[_vestingId] = vesting_;
+
+        emit VestingCreated(_vestingId);
+
+        return _vestingId;
+    }
+
+    // withdraw tokens from vesting and transfer to beneficiary
+    function _withdrawFromVesting(
+        uint256 vestingId_
+    ) internal virtual returns (uint256 _amountToPay, address _token) {
+        VestingData storage _vesting = _vestings[vestingId_];
+
+        require(
+            msg.sender == _vesting.beneficiary,
+            "VestingWallet: only befeciary can withdraw from his vesting"
+        );
+
+        _amountToPay = getWithdrawableAmount(vestingId_);
+        _token = _vesting.vestingToken;
+
+        _vesting.paidAmount += _amountToPay;
+
+        emit WithdrawFromVesting(vestingId_, _amountToPay);
+    }
+
+    // default implementation of vesting calculation
+    function _vestingCalculation(
+        Schedule memory schedule_,
+        uint256 totalVestingAmount_,
+        uint256 vestingStartTime_,
+        uint256 timestampUpTo_,
+        uint256 timestampCurrent_
```

Comment:
But do we actually need 2 exactly the same variables here?

Reply:
actually no